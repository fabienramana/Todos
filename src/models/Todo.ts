export default interface Todo{
    content: string,
    completed: boolean
}