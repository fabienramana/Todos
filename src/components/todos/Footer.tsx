import Todo from "../../models/Todo"

type AppProps = {
    todoArray: Todo[],
    removeCompletedTodos: ()=> void,
    filter: string,
    setFilter: (filter: string) => void
}

export default function Footer({todoArray, removeCompletedTodos, filter, setFilter}: AppProps): JSX.Element {

    const nbrTodosLeft = todoArray.filter((todo) => todo.completed === false).length

    return(
        <footer className="footer">
            <span className="todo-count">{nbrTodosLeft} item{nbrTodosLeft !== 1 && "s"} left</span>
            <ul className="filters">
                <li>
                    <a className={filter==="all" ? "selected" : ""} onClick={() => setFilter("all")}  href="#/">All</a>
                </li>
                <li>
                    <a className={filter==="active" ? "selected" : ""} onClick={() => setFilter("active")} href="#/active">Active</a>
                </li>
                <li>
                    <a className={filter==="completed" ? "selected" : ""} onClick={() => setFilter("completed")} href="#/completed">Completed</a>
                </li>
            </ul>
            <button className="clear-completed" onClick={() => removeCompletedTodos()}>Clear completed</button>
        </footer>
    )
}