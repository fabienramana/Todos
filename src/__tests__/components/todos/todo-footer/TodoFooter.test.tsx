import { render, screen, fireEvent } from "@testing-library/react";
import Footer from "../../../../components/todos/todo-footer/TodoFooter";
import Todo from "../../../../models/Todo";


describe('Footer component', ()=>{

    const listOfTodos: Array<Todo> = [
      {
          id:1,
          content:"Wake up early",
          completed: false
      },
      {
          id:2,
          content:"Sleep early",
          completed: false
      },
      {
          id:3,
          content:"Do the chores",
          completed: true
      } 
  
  ]
  
    let filter = "all"
  
    /****************** Comment faire pour matcher le texte quand split en plusieurs balises ? EX : <p><strong>0</strong> item left</p> ****************/   
    test('should render number of todos left', ()=>{
  
      const removeCompletedTodos = () => {};
      const setFilter = () => {};
  
      render(<Footer todoArray={listOfTodos} 
              removeCompletedTodos={removeCompletedTodos}
              setFilter={setFilter}
              filter={filter}/>)
  
      const nbrTodosLeft = listOfTodos.filter((todo)=> todo.completed === false).length
  
      expect(screen.getByText(`${nbrTodosLeft} items left`)).toBeInTheDocument();
    })
  
    test("should call removeCompletedTodos if Clear completed button is clicked", ()=>{
  
      const removeCompletedTodos = jest.fn();
      const setFilter = () => {};
  
      render(<Footer todoArray={listOfTodos} 
        removeCompletedTodos={removeCompletedTodos}
        setFilter={setFilter}
        filter={filter}/>)
  
        fireEvent.click(screen.getByRole('button'));
  
        expect(removeCompletedTodos).toBeCalledTimes(1);
    })
  
    test("should call setFilter with all if all filter is clicked", ()=>{
  
      const removeCompletedTodos = jest.fn();
      const setFilter = jest.fn();
  
      render(<Footer todoArray={listOfTodos} 
        removeCompletedTodos={removeCompletedTodos}
        setFilter={setFilter}
        filter={filter}/>)
  
        fireEvent.click(screen.getByRole('link', {
          name: /all/i
        }));
  
        expect(setFilter).toBeCalledTimes(1);
        expect(setFilter).toBeCalledWith(filter)
    })
  
    test("should call setFilter with active if active filter is clicked", ()=>{
      
      const removeCompletedTodos = () => {};
      const setFilter = jest.fn();
  
      filter = "active"
      render(<Footer todoArray={listOfTodos} 
        removeCompletedTodos={removeCompletedTodos}
        setFilter={setFilter}
        filter={filter}/>)
  
        fireEvent.click(screen.getByRole('link', {
          name: /active/i
        }));
  
        expect(setFilter).toBeCalledTimes(1);
        expect(setFilter).toBeCalledWith(filter)
    })
  
    test("should call setFilter with completed if active filter is clicked", ()=>{
  
      const removeCompletedTodos = () => {};
      const setFilter = jest.fn();
  
      filter= "completed"
  
      render(<Footer todoArray={listOfTodos} 
        removeCompletedTodos={removeCompletedTodos}
        setFilter={setFilter}
        filter={filter}/>)
  
        fireEvent.click(screen.getByRole('link', {
          name: /completed/i
        }));
  
        expect(setFilter).toBeCalledTimes(1);
        expect(setFilter).toBeCalledWith(filter)
    })
  })
  