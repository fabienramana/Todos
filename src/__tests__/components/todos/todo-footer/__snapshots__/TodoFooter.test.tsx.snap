// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer component should match snapshot 1`] = `
<DocumentFragment>
  <footer
    class="footer"
  >
    <span
      class="todo-count"
    >
      3 items left
    </span>
    <ul
      class="filters"
    >
      <li>
        <a
          aria-current="page"
          class="selected"
          data-testid="all"
          href="/"
        >
          All
        </a>
      </li>
      <li>
        <a
          data-testid="active"
          href="/active"
        >
          Active
        </a>
      </li>
      <li>
        <a
          data-testid="completed"
          href="/completed"
        >
          Completed
        </a>
      </li>
    </ul>
    <button
      class="clear-completed"
    >
      Clear completed
    </button>
  </footer>
</DocumentFragment>
`;
